src/transform/ancestor.js

```diff
@@ -118,7 +118,7 @@ function canBeLifted(doc, range) {
   }
 }
 
-// :: (Node, Pos, ?Pos) → bool
+// :: (Node, Pos, ?Pos) → ?{found: any, range: any}
```

Comment:
> Most of the hacking-around happens in handleBadLines

Most of these lines don't look that bad :). I do guess the end conclusion of this experiment is going to be me giving up on statically typed JS for another few years, though.